### PR TITLE
#25 recognize @link when class name is missing

### PR DESCRIPTION
**Pull request by bbottema**

### CODE REVIEW

File: therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParser.java [hunk 1]
```diff
@@ -63,7 +63,7 @@ private static InlineLink createLinkElement(String value) {
         String classRef = ref[0];
         String memberRef = ref.length > 1 ? ref[1] : null;
 
-        Link link = new Link(label, classRef, memberRef);
+        Link link = new Link(label, classRef.isEmpty() ? null : classRef, memberRef);
```
Comment (dnault) [inline therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParser.java]: I wonder if there's a way to infer the class name from the enclosing elements?

**Comment by dnault** [comment]
Thank you for this!

File: therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParser.java (hunk 1 above)
Comment (bbottema) [reply therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/internal/parser/CommentParser.java]: I'm now passing along the current class as fall-back. It's a little verbose passing it around to all the methods, but it works I guess.

File: therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/Link.java [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
+import java.util.Arrays;
 import java.util.Objects;
 
 public class Link {
     private final String label;
     private final String referencedClassName;
     private final String referencedMemberName;
-
-    public Link(String label, String referencedClassName, String referencedMemberName) {
+    private final String[] params;
```
Comment (dnault) [inline therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/Link.java]: To maintain immutability (and consistency with other classes) can you please change this to a `List<String>` and initialize it with `unmodifiableDefensiveCopy`?

**Review by dnault — CHANGES_REQUESTED** [review]

File: therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/Link.java (hunk 2 above)
Comment (bbottema) [reply therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/Link.java]: Done

**Comment by joffrey-bion** [comment]
Is this the actual Javadoc tool behavior?

File: therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/Link.java [hunk 3]
```diff
@@ -49,9 +59,20 @@ public String toString() {
...
             sb.append('#').append(referencedMemberName);
+    
+            if (params != null && !params.isEmpty()) {
+                sb.append('(');
+                for (int i = 0; i < params.size(); i++) {
+                    sb.append(params.get(i));
+                    if (i < params.size() - 1) {
+                        sb.append(", ");
+                    }
+                }
```
Comment (joffrey-bion) [inline therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/Link.java:73]: It looks like this calls for `Collectors.joining(",")` :)

**Comment by bbottema** [comment]
> Is this the actual Javadoc tool behavior?

I don't understand the question.

**Comment by joffrey-bion** [comment]
I mean, does the Javadoc tool consider @link valid when not followed by a reference? Does it fallback to anything? Does it display a literal "@link" without href? Nothing at all? I believe this is what should drive therapi-runtime-javadoc's behavior

**Comment by bbottema** [comment]
> I mean, does the Javadoc tool consider @link valid when not followed by a reference? Does it fallback to anything? Does it display a literal "@link" without href? Nothing at all? I believe this is what should drive therapi-runtime-javadoc's behavior

That's more a question for @dnault then. I just fixed support for certain valid constructions.

**Review by joffrey-bion — COMMENTED** [review]
> > I mean, does the Javadoc tool consider @link valid when not followed by a reference? Does it fallback to anything? Does it display a literal "@link" without href? Nothing at all? I believe this is what should drive therapi-runtime-javadoc's behavior
> 
> That's more a question for @dnault then. I just fixed support for certain valid constructions.

That's right, my bad. I read the title too fast and thought you added support for empty content of `link` tag. But in fact, you added support for member references without class names.

Great work btw, thanks for this PR.

File: therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/Link.java (hunk 3 above)
Comment (bbottema) [reply therapi-runtime-javadoc/src/main/java/com/github/therapi/runtimejavadoc/Link.java:73]: Sorry, not in JDK 7 :/

**Comment by bbottema** [comment]
Fixed a small new bug that was introduced and added test cases for that. Think I covered everything now. Can we finish this pull request, @dnault?

**Pull request merged**